README.md
[![Build Status](https://travis-ci.org/panda3d/panda3d.svg?branch=master)](https://travis-ci.org/panda3d/panda3d)

<img src="https://avatars2.githubusercontent.com/u/590956?v=3&s=500" align="right" width="200"/>

Panda3D - Miles of 1.11.0!
=======

Panda3D is a game engine, a framework for 3D rendering and game development for
Python and C++ programs.  Panda3D is open-source and free for any purpose,
including commercial ventures, thanks to its
[liberal license](https://www.panda3d.org/license/). To learn more about
Panda3D's capabilities, visit the [gallery](https://www.panda3d.org/gallery/)
and the [feature list](https://www.panda3d.org/features/).  To learn how to
use Panda3D, check the [documentation](https://www.panda3d.org/documentation/)
resources. If you get stuck, ask for help from our active
[community](https://discourse.panda3d.org).

Panda3D is licensed under the Modified BSD License.  See the LICENSE file for
more details.

Installing Panda3D Miles
==================

The latest Panda3D can be downloaded from
[this page](https://dl.broadwaypixels.com/Panda3D-1.11.0.exe).
If you are familiar with installing Python packages, you can use
the following comand:

```bash
pip install panda3d
```

The easiest way to install the latest development build of Panda3D
into an existing Python installation is using the following command:

```bash
pip install --pre --extra-index-url https://archive.panda3d.org/ panda3d
```

If this command fails, please make sure your version of pip is up-to-date.

If you prefer to install the full SDK with all tools, the latest development
builds can be obtained from [this page](https://www.panda3d.org/download.php?version=devel&sdk).

These are automatically kept up-to-date with the latest GitHub version of Panda.

Building Panda3D Miles
================

Windows
-------

Should be the same, if not let me know. :)

You can build Panda3D with the Microsoft Visual C++ 2015 or 2017 compiler,
which can be downloaded for free from the [Visual Studio site](https://visualstudio.microsoft.com/downloads/).
You will also need to install the [Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk),
and if you intend to target Windows XP, you will also need the
[Windows 7.1 SDK](https://www.microsoft.com/en-us/download/details.aspx?id=8279).

You will also need to have the third-party dependency libraries available for
the build scripts to use. After acquiring these dependencies, you may simply build Panda3D from the
command prompt using the following command.  (Change `14.1` to `14` if you are
using Visual C++ 2015 instead of 2017.  Add the `--windows-sdk=10` option if
you don't need to support Windows XP and did not install the Windows 7.1 SDK.)

```bash
makepanda\makepanda.bat --everything --installer --msvc-version=14.1 --no-eigen --threads=2
```

When the build succeeds, it will produce an .exe file that you can use to
install Panda3D on your system.

Note: you may choose to remove --no-eigen and build with Eigen support in
order to improve runtime performance.  However, this will cause the build to
take hours to complete, as Eigen is a heavily template-based library, and the
the MSVC compiler does not perform well under these circumstances.

Linux
-----

Building Panda3D on Linux is easy.  All you need is to invoke the makepanda
script using the version of Python that you want Panda3D to be built against.

Run makepanda.py with the --help option to see which options are available.
Usually, you will want to specify the --everything option (which builds with
support for all features for which it detects the prerequisite dependencies)
and the --installer option (which produces an installable .deb or .rpm file
for you to install, depending on your distribution).

The following command illustrates how to build Panda3D with some common
options:
```bash
python makepanda/makepanda.py --everything --installer --no-egl --no-gles --no-gles2 --no-opencv
```

You will probably see some warnings saying that it's unable to find several
dependency packages.  You should determine which ones you want to include in
your build and install the respective development packages.  You may visit
[this manual page](https://www.panda3d.org/manual/?title=Third-party_dependencies_and_license_info)
for an overview of the various dependencies.

If you are on Ubuntu, this command should cover the most frequently
used third-party packages:

```bash
sudo apt-get install build-essential pkg-config fakeroot python-dev libpng-dev libjpeg-dev libtiff-dev zlib1g-dev libssl-dev libx11-dev libgl1-mesa-dev libxrandr-dev libxxf86dga-dev libxcursor-dev bison flex libfreetype6-dev libvorbis-dev libeigen3-dev libopenal-dev libode-dev libbullet-dev nvidia-cg-toolkit libgtk2.0-dev libassimp-dev libopenexr-dev
```

Once Panda3D has built, you can either install the .deb or .rpm package that
it produced, depending on which Linux distribution you are using.  For example,
to install the package on Debian or Ubuntu, use this:

```bash
sudo dpkg -i panda3d*.deb
```

If you are not using a Linux distribution that supports .deb or .rpm packages, you
may have to use the installpanda.py script instead, which will directly copy the
files into the appropriate locations on your computer.  You may have to run the
`ldconfig` tool in order to update your library cache after installing Panda3D.

Alternatively, you can add the `--wheel` option, which will produce a .whl
file that can be installed into a Python installation using `pip`.

macOS
-----

On macOS, you will need to download a set of precompiled thirdparty packages in order to
compile Panda3D.

After placing the thirdparty directory inside the panda3d source directory,
you may build Panda3D using a command like the following:

```bash
python makepanda/makepanda.py --everything --installer
```

You may target a specific minimum macOS version using the --osxtarget flag
followed by the release number, eg. 10.7 or 10.9.

If the build was successful, makepanda will have generated a .dmg file in
the source directory containing the installer.  Simply open it and run the
package file in order to install the SDK onto your system.

FreeBSD
-------

Building on FreeBSD is very similar to building on Linux.  You will need to
install the requisite packages using the system package manager.  To install
the recommended set of dependencies, you can use this command:

```bash
pkg install pkgconf png jpeg-turbo tiff freetype2 eigen squish openal opusfile libvorbis libX11 libGL ode bullet assimp openexr
```

You will also need to choose which version of Python you want to use.
Install the appropriate package for it (such as `python2` or `python36`) and
run the makepanda script with your chosen Python version:

```bash
python3.6 makepanda/makepanda.py --everything --installer --no-egl --no-gles --no-gles2
```

If successful, this will produce a .pkg file in the root of the source
directory which you can install using `pkg install`.

Android
-------

Note: building on Android is very experimental and not guaranteed to work.

You can experimentally build the Android Python runner via the [termux](https://termux.com/)
shell.  You will need to install [Termux](https://play.google.com/store/apps/details?id=com.termux)
and [Termux API](https://play.google.com/store/apps/details?id=com.termux.api)
from the Play Store.  Many of the dependencies can be installed by running the
following command in the Termux shell:

```bash
pkg install python-dev termux-tools ndk-stl ndk-sysroot clang libvorbis-dev libopus-dev opusfile-dev openal-soft-dev freetype-dev harfbuzz-dev libpng-dev ecj4.6 dx patchelf aapt apksigner libcrypt-dev
```

Then, you can build and install the .apk right away using these commands:

```bash
python makepanda/makepanda.py --everything --target android-21 --installer
xdg-open panda3d.apk
```

To launch a Python program from Termux, you can use the `run_python.sh` script
inside the `panda/src/android` directory.  It will launch Python in a separate
activity, load it with the Python script you passed as argument, and use a
socket for returning the command-line output to the Termux shell.  Do note
that this requires the Python application to reside on the SD card and that
Termux needs to be set up with access to the SD card (using the
`termux-setup-storage` command).

Running Tests
=============

Install [PyTest](https://docs.pytest.org/en/latest/getting-started.html#installation)
and run the `pytest` command.  If you have not installed Panda3D, you will
need to configure your enviroment by pointing the `PYTHONPATH` variable at
the `built` directory.  On Linux, you will also need to point the
`LD_LIBRARY_PATH` variable at the `built/lib` directory.

As a convenience, you can alternatively pass the `--tests` option to makepanda.

Reporting Issues
================

If you encounter any bugs when using Panda3D, please report them in the bug
tracker.

Make sure to first use the search function to see if the bug has already been
reported.  When filling out a bug report, make sure that you include as much
information as possible to help the developers track down the issue, such as
your version of Panda3D, operating system, architecture, and any code and
models that are necessary for the developers to reproduce the issue.

If you're not sure whether you've encountered a bug, feel free to ask about
it in the forums or the IRC channel first.


ReleaseNotes
------------------------  RELEASE 1.10.1  -----------------------

* Fix regression with Audio3DManager.setSoundVelocityAuto()
* Audio3DManager accepts tuple in setSoundVelocity/setListenerVelocity

------------------------  RELEASE 1.10.0  -----------------------

This is a major release with significant changes.  Please review the
changes when upgrading.  The list below is by no means exhaustive, but
should contain the most important changes.

General
* Experimental ability to build for Android
* New input framework to natively support gamepads, joysticks, etc.
* Multi-threaded render pipeline is a lot more stable now
* New setuptools-based deployment pipeline
* Improvements to mouselook smoothness
* Cache is now at $XDG_CACHE_HOME/panda3d (~/.cache/panda3d), not ~/.panda3d
* Addition of unit test suite
* Many improvements to thread safety
* Many performance improvements
* Tons of bugfixes
* Big style cleanup of C++ source code

Python API
* Complete support for Python 3
* Support for coroutines and async/await
* Property interfaces have been added for many settings
* More flexible handling for keyboard arguments in C++ APIs
* Python bindings are completely separated out of the C++ libraries.
* Interrogate binding generator has many improvements.
* Use of pandac.PandaModules is discouraged, use panda3d.core et al
* Use of libRocket is discouraged due to lack of Python 3 support
* Tasks are now sorted in addition order when lacking a sort value
* Fixes iris/fade transitions for extreme aspect ratios
* WeakNodePath is now exposed to Python
* WindowProperties.size(x, y) deprecated; use WindowProperties(size=(x, y))
* Calling bare run() is deprecated, use base.run() instead
* downcastTo*() methods have been removed, they were already no-ops

Rendering
* Add new shader-based terrain rendering method (ShaderTerrainMesh)
* The default ColorAttrib mode is now T_vertex
* The ColorAttrib T_off mode now properly disables vertex colors entirely
* Make handling of color attributes more consistent between renderers
* Ability to create an OpenGL core profile context; set "gl-version 3 2"
* Experimental support for reverse-Z rendering for best depth precision
* sRGB framebuffers supported more widely
* Support for infinite near/far clip in lens
* Add some PBR material parameters to material class
* Addition of more built-in GLSL shader inputs; see manual.
* Add p3d_FragData[] GLSL output for MRT in GLSL 1.30
* Add flag enabling vertex shader control over point size
* Support signed ints and double-precision floats in vertex data with GLSL
* Support unsigned 11/10/10-bit floating-point textures and vertex data
* Support for SSBOs via ShaderBuffer class
* Support OpenGL FBO buffers without any attachments
* Support passing uint variables to GLSL shader
* Allow rendering objects with empty vertex data (for vertex pulling)
* Add LogicOpAttrib, for supporting logical operator blending
* Improvements to OpenGL ES support
* Support for geometry with adjacency information
* Change default alpha blending to improve blending rendered result
* New method for obtaining native OpenGL texture object
* Support windowless offscreen rendering on macOS
* Panda resets OpenGL state better before and after draw callbacks
* OpenGL renderer better supports debugging tools like apitrace
* Support fixed-depth billboards, useful for 2D tags that don't change size

Shader generator
* Significant performance improvements
* Support for point light shadows
* Hardware skinning support
* Changes to match fixed-function pipeline better
* Fixes for normal vector normalization
* Support multiple normal maps (uses Reoriented Normal Mapping)
* Tracks modifications to materials and texture stages automatically

Lighting
* Allow specifying light color based on color temperature
* Setting specular color of a light separately is deprecated
* New GLSL inputs to make implementing lighting in shaders much easier
* Add representation for sphere light and rectangle light
* Efficiency improvements for passing light information to shader
* Interocular distance for shadow cameras now always defaults to 0
* Add low-level lighting module from RenderPipeline

Textures
* Support cube map arrays
* Support buffer textures
* Many more texture formats supported
* BC4 and BC5 compression modes supported
* Proper depth textures supported in DirectX 9 renderer
* set_ram_image(_as) directly supports buffer protocol
* TexturePeeker supports more formats and component types

Text
* Dramatic improvements to text rendering performance
* Support for HarfBuzz for higher-quality text shaping and kerning
* Support for right-to-left text
* Support for signed-distance-field rendering in egg-mkfont

Audio/video
* The default unit for audio is now 1 meter for each Panda unit.
* Native .flac loader
* Support videos with alpha channel in ffmpeg
* OpenAL stability improvements, especially on macOS
* Support loading .opus files with libopusfile
* Fix various memory leaks

Physics / collisions
* CollisionTube is renamed to CollisionCapsule.
* Box-box collision test is improved to work well with the Pusher
* More box tests for collision system: box-into-plane, box-into-poly
* Capsule (tube) can be used as "from" shape into plane, sphere, capsule, box
* Bullet objects are serializable to .bam files.
* Bullet bindings are now thread safe.
* Bullet debug drawer is more efficient; no longer inherits GeomNode.
* Various fixes to bullet vehicle wheel synchronization
* PhysX bindings are deprecated.

Pipeline / loading
* Support for Assimp library to load a broad variety of model formats
* Ability to specify min-lod, max-lod, lod-bias in .egg file
* Egg file materials support PBR-style material parameterization

include/dcParser.h
/* A Bison parser, made by GNU Bison 2.7.  */

/* Bison interface for Yacc-like parsers in C
   
      Copyright (C) 1984, 1989-1990, 2000-2012 Free Software Foundation, Inc.
   
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* As a special exception, you may create a larger work that contains
   part or all of the Bison parser skeleton and distribute that work
   under terms of your choice, so long as that work isn't itself a
   parser generator using the skeleton or a modified version thereof
   as a parser skeleton.  Alternatively, if you modify or redistribute
   the parser skeleton itself, you may (at your option) remove this
   special exception, which will cause the skeleton and the resulting
   Bison output files to be licensed under the GNU General Public
   License without this special exception.
   
   This special exception was added by the Free Software Foundation in
   version 2.2 of Bison.  */

#ifndef YY_DCYY_BUILT_TMP_DCPARSER_YXX_H_INCLUDED
# define YY_DCYY_BUILT_TMP_DCPARSER_YXX_H_INCLUDED
/* Enabling traces.  */
#ifndef YYDEBUG
# define YYDEBUG 0
#endif
#if YYDEBUG
extern int dcyydebug;
#endif

/* Tokens.  */
#ifndef YYTOKENTYPE
# define YYTOKENTYPE
   /* Put the tokens into the symbol table, so that GDB and other debuggers
      know about them.  */
   enum yytokentype {
     UNSIGNED_INTEGER = 258,
     SIGNED_INTEGER = 259,
     REAL = 260,
     STRING = 261,
     IDENTIFIER = 262,
     HEX_STRING = 263,
     KEYWORD = 264,
     KW_DCLASS = 265,
     KW_STRUCT = 266,
     KW_FROM = 267,
     KW_IMPORT = 268,
     KW_TYPEDEF = 269,
     KW_KEYWORD = 270,
     KW_SWITCH = 271,
     KW_CASE = 272,
     KW_DEFAULT = 273,
     KW_BREAK = 274,
     KW_INT8 = 275,
     KW_INT16 = 276,
     KW_INT32 = 277,
     KW_INT64 = 278,
     KW_UINT8 = 279,
     KW_UINT16 = 280,
     KW_UINT32 = 281,
     KW_UINT64 = 282,
     KW_FLOAT64 = 283,
     KW_STRING = 284,
     KW_BLOB = 285,
     KW_BLOB32 = 286,
     KW_INT8ARRAY = 287,
     KW_INT16ARRAY = 288,
     KW_INT32ARRAY = 289,
     KW_UINT8ARRAY = 290,
     KW_UINT16ARRAY = 291,
     KW_UINT32ARRAY = 292,
     KW_UINT32UINT8ARRAY = 293,
     KW_CHAR = 294,
     START_DC = 295,
     START_PARAMETER_VALUE = 296,
     START_PARAMETER_DESCRIPTION = 297
   };
#endif
/* Tokens.  */
#define UNSIGNED_INTEGER 258
#define SIGNED_INTEGER 259
#define REAL 260
#define STRING 261
#define IDENTIFIER 262
#define HEX_STRING 263
#define KEYWORD 264
#define KW_DCLASS 265
#define KW_STRUCT 266
#define KW_FROM 267
#define KW_IMPORT 268
#define KW_TYPEDEF 269
#define KW_KEYWORD 270
#define KW_SWITCH 271
#define KW_CASE 272
#define KW_DEFAULT 273
#define KW_BREAK 274
#define KW_INT8 275
#define KW_INT16 276
#define KW_INT32 277
#define KW_INT64 278
#define KW_UINT8 279
#define KW_UINT16 280
#define KW_UINT32 281
#define KW_UINT64 282
#define KW_FLOAT64 283
#define KW_STRING 284
#define KW_BLOB 285
#define KW_BLOB32 286
#define KW_INT8ARRAY 287
#define KW_INT16ARRAY 288
#define KW_INT32ARRAY 289
#define KW_UINT8ARRAY 290
#define KW_UINT16ARRAY 291
#define KW_UINT32ARRAY 292
#define KW_UINT32UINT8ARRAY 293
#define KW_CHAR 294
#define START_DC 295
#define START_PARAMETER_VALUE 296
#define START_PARAMETER_DESCRIPTION 297



#if ! defined YYSTYPE && ! defined YYSTYPE_IS_DECLARED

# define yystype YYSTYPE /* obsolescent; will be withdrawn */
# define YYSTYPE_IS_DECLARED 1
#endif

extern YYSTYPE dcyylval;

#ifdef YYPARSE_PARAM
#if defined __STDC__ || defined __cplusplus
int dcyyparse (void *YYPARSE_PARAM);
#else
int dcyyparse ();
#endif
#else /* ! YYPARSE_PARAM */
#if defined __STDC__ || defined __cplusplus
int dcyyparse (void);
#else
int dcyyparse ();
#endif
#endif /* ! YYPARSE_PARAM */

#endif /* !YY_DCYY_BUILT_TMP_DCPARSER_YXX_H_INCLUDED  */


include/cppBison.h
/* A Bison parser, made by GNU Bison 2.7.  */

/* Bison interface for Yacc-like parsers in C
   
      Copyright (C) 1984, 1989-1990, 2000-2012 Free Software Foundation, Inc.
   
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* As a special exception, you may create a larger work that contains
   part or all of the Bison parser skeleton and distribute that work
   under terms of your choice, so long as that work isn't itself a
   parser generator using the skeleton or a modified version thereof
   as a parser skeleton.  Alternatively, if you modify or redistribute
   the parser skeleton itself, you may (at your option) remove this
   special exception, which will cause the skeleton and the resulting
   Bison output files to be licensed under the GNU General Public
   License without this special exception.
   
   This special exception was added by the Free Software Foundation in
   version 2.2 of Bison.  */

#ifndef YY_CPPYY_BUILT_TMP_CPPBISON_YXX_H_INCLUDED
# define YY_CPPYY_BUILT_TMP_CPPBISON_YXX_H_INCLUDED
/* Enabling traces.  */
#ifndef YYDEBUG
# define YYDEBUG 0
#endif
#if YYDEBUG
extern int cppyydebug;
#endif

/* Tokens.  */
#ifndef YYTOKENTYPE
# define YYTOKENTYPE
   /* Put the tokens into the symbol table, so that GDB and other debuggers
      know about them.  */
   enum yytokentype {
     REAL = 258,
     INTEGER = 259,
     CHAR_TOK = 260,
     SIMPLE_STRING = 261,
     SIMPLE_IDENTIFIER = 262,
     STRING_LITERAL = 263,
     CUSTOM_LITERAL = 264,
     IDENTIFIER = 265,
     TYPENAME_IDENTIFIER = 266,
     TYPEPACK_IDENTIFIER = 267,
     SCOPING = 268,
     TYPEDEFNAME = 269,
     ELLIPSIS = 270,
     OROR = 271,
     ANDAND = 272,
     EQCOMPARE = 273,
     NECOMPARE = 274,
     LECOMPARE = 275,
     GECOMPARE = 276,
     LSHIFT = 277,
     RSHIFT = 278,
     POINTSAT_STAR = 279,
     DOT_STAR = 280,
     UNARY = 281,
     UNARY_NOT = 282,
     UNARY_NEGATE = 283,
     UNARY_MINUS = 284,
     UNARY_PLUS = 285,
     UNARY_STAR = 286,
     UNARY_REF = 287,
     POINTSAT = 288,
     SCOPE = 289,
     PLUSPLUS = 290,
     MINUSMINUS = 291,
     TIMESEQUAL = 292,
     DIVIDEEQUAL = 293,
     MODEQUAL = 294,
     PLUSEQUAL = 295,
     MINUSEQUAL = 296,
     OREQUAL = 297,
     ANDEQUAL = 298,
     XOREQUAL = 299,
     LSHIFTEQUAL = 300,
     RSHIFTEQUAL = 301,
     ATTR_LEFT = 302,
     ATTR_RIGHT = 303,
     KW_ALIGNAS = 304,
     KW_ALIGNOF = 305,
     KW_AUTO = 306,
     KW_BEGIN_PUBLISH = 307,
     KW_BLOCKING = 308,
     KW_BOOL = 309,
     KW_CATCH = 310,
     KW_CHAR = 311,
     KW_CHAR16_T = 312,
     KW_CHAR32_T = 313,
     KW_CLASS = 314,
     KW_CONST = 315,
     KW_CONSTEXPR = 316,
     KW_CONST_CAST = 317,
     KW_DECLTYPE = 318,
     KW_DEFAULT = 319,
     KW_DELETE = 320,
     KW_DOUBLE = 321,
     KW_DYNAMIC_CAST = 322,
     KW_ELSE = 323,
     KW_END_PUBLISH = 324,
     KW_ENUM = 325,
     KW_EXTENSION = 326,
     KW_EXTERN = 327,
     KW_EXPLICIT = 328,
     KW_PUBLISHED = 329,
     KW_FALSE = 330,
     KW_FINAL = 331,
     KW_FLOAT = 332,
     KW_FRIEND = 333,
     KW_FOR = 334,
     KW_GOTO = 335,
     KW_HAS_VIRTUAL_DESTRUCTOR = 336,
     KW_IF = 337,
     KW_INLINE = 338,
     KW_INT = 339,
     KW_IS_ABSTRACT = 340,
     KW_IS_BASE_OF = 341,
     KW_IS_CLASS = 342,
     KW_IS_CONSTRUCTIBLE = 343,
     KW_IS_CONVERTIBLE_TO = 344,
     KW_IS_DESTRUCTIBLE = 345,
     KW_IS_EMPTY = 346,
     KW_IS_ENUM = 347,
     KW_IS_FINAL = 348,
     KW_IS_FUNDAMENTAL = 349,
     KW_IS_POD = 350,
     KW_IS_POLYMORPHIC = 351,
     KW_IS_STANDARD_LAYOUT = 352,
     KW_IS_TRIVIAL = 353,
     KW_IS_UNION = 354,
     KW_LONG = 355,
     KW_MAKE_MAP_KEYS_SEQ = 356,
     KW_MAKE_MAP_PROPERTY = 357,
     KW_MAKE_PROPERTY = 358,
     KW_MAKE_PROPERTY2 = 359,
     KW_MAKE_SEQ = 360,
     KW_MAKE_SEQ_PROPERTY = 361,
     KW_MUTABLE = 362,
     KW_NAMESPACE = 363,
     KW_NEW = 364,
     KW_NOEXCEPT = 365,
     KW_NULLPTR = 366,
     KW_OPERATOR = 367,
     KW_OVERRIDE = 368,
     KW_PRIVATE = 369,
     KW_PROTECTED = 370,
     KW_PUBLIC = 371,
     KW_REGISTER = 372,
     KW_REINTERPRET_CAST = 373,
     KW_RETURN = 374,
     KW_SHORT = 375,
     KW_SIGNED = 376,
     KW_SIZEOF = 377,
     KW_STATIC = 378,
     KW_STATIC_ASSERT = 379,
     KW_STATIC_CAST = 380,
     KW_STRUCT = 381,
     KW_TEMPLATE = 382,
     KW_THREAD_LOCAL = 383,
     KW_THROW = 384,
     KW_TRUE = 385,
     KW_TRY = 386,
     KW_TYPEDEF = 387,
     KW_TYPEID = 388,
     KW_TYPENAME = 389,
     KW_UNDERLYING_TYPE = 390,
     KW_UNION = 391,
     KW_UNSIGNED = 392,
     KW_USING = 393,
     KW_VIRTUAL = 394,
     KW_VOID = 395,
     KW_VOLATILE = 396,
     KW_WCHAR_T = 397,
     KW_WHILE = 398,
     START_CPP = 399,
     START_CONST_EXPR = 400,
     START_TYPE = 401
   };
#endif
/* Tokens.  */
#define REAL 258
#define INTEGER 259
#define CHAR_TOK 260
#define SIMPLE_STRING 261
#define SIMPLE_IDENTIFIER 262
#define STRING_LITERAL 263
#define CUSTOM_LITERAL 264
#define IDENTIFIER 265
#define TYPENAME_IDENTIFIER 266
#define TYPEPACK_IDENTIFIER 267
#define SCOPING 268
#define TYPEDEFNAME 269
#define ELLIPSIS 270
#define OROR 271
#define ANDAND 272
#define EQCOMPARE 273
#define NECOMPARE 274
#define LECOMPARE 275
#define GECOMPARE 276
#define LSHIFT 277
#define RSHIFT 278
#define POINTSAT_STAR 279
#define DOT_STAR 280
#define UNARY 281
#define UNARY_NOT 282
#define UNARY_NEGATE 283
#define UNARY_MINUS 284
#define UNARY_PLUS 285
#define UNARY_STAR 286
#define UNARY_REF 287
#define POINTSAT 288
#define SCOPE 289
#define PLUSPLUS 290
#define MINUSMINUS 291
#define TIMESEQUAL 292
#define DIVIDEEQUAL 293
#define MODEQUAL 294
#define PLUSEQUAL 295
#define MINUSEQUAL 296
#define OREQUAL 297
#define ANDEQUAL 298
#define XOREQUAL 299
#define LSHIFTEQUAL 300
#define RSHIFTEQUAL 301
#define ATTR_LEFT 302
#define ATTR_RIGHT 303
#define KW_ALIGNAS 304
#define KW_ALIGNOF 305
#define KW_AUTO 306
#define KW_BEGIN_PUBLISH 307
#define KW_BLOCKING 308
#define KW_BOOL 309
#define KW_CATCH 310
#define KW_CHAR 311
#define KW_CHAR16_T 312
#define KW_CHAR32_T 313
#define KW_CLASS 314
#define KW_CONST 315
#define KW_CONSTEXPR 316
#define KW_CONST_CAST 317
#define KW_DECLTYPE 318
#define KW_DEFAULT 319
#define KW_DELETE 320
#define KW_DOUBLE 321
#define KW_DYNAMIC_CAST 322
#define KW_ELSE 323
#define KW_END_PUBLISH 324
#define KW_ENUM 325
#define KW_EXTENSION 326
#define KW_EXTERN 327
#define KW_EXPLICIT 328
#define KW_PUBLISHED 329
#define KW_FALSE 330
#define KW_FINAL 331
#define KW_FLOAT 332
#define KW_FRIEND 333
#define KW_FOR 334
#define KW_GOTO 335
#define KW_HAS_VIRTUAL_DESTRUCTOR 336
#define KW_IF 337
#define KW_INLINE 338
#define KW_INT 339
#define KW_IS_ABSTRACT 340
#define KW_IS_BASE_OF 341
#define KW_IS_CLASS 342
#define KW_IS_CONSTRUCTIBLE 343
#define KW_IS_CONVERTIBLE_TO 344
#define KW_IS_DESTRUCTIBLE 345
#define KW_IS_EMPTY 346
#define KW_IS_ENUM 347
#define KW_IS_FINAL 348
#define KW_IS_FUNDAMENTAL 349
#define KW_IS_POD 350
#define KW_IS_POLYMORPHIC 351
#define KW_IS_STANDARD_LAYOUT 352
#define KW_IS_TRIVIAL 353
#define KW_IS_UNION 354
#define KW_LONG 355
#define KW_MAKE_MAP_KEYS_SEQ 356
#define KW_MAKE_MAP_PROPERTY 357
#define KW_MAKE_PROPERTY 358
#define KW_MAKE_PROPERTY2 359
#define KW_MAKE_SEQ 360
#define KW_MAKE_SEQ_PROPERTY 361
#define KW_MUTABLE 362
#define KW_NAMESPACE 363
#define KW_NEW 364
#define KW_NOEXCEPT 365
#define KW_NULLPTR 366
#define KW_OPERATOR 367
#define KW_OVERRIDE 368
#define KW_PRIVATE 369
#define KW_PROTECTED 370
#define KW_PUBLIC 371
#define KW_REGISTER 372
#define KW_REINTERPRET_CAST 373
#define KW_RETURN 374
#define KW_SHORT 375
#define KW_SIGNED 376
#define KW_SIZEOF 377
#define KW_STATIC 378
#define KW_STATIC_ASSERT 379
#define KW_STATIC_CAST 380
#define KW_STRUCT 381
#define KW_TEMPLATE 382
#define KW_THREAD_LOCAL 383
#define KW_THROW 384
#define KW_TRUE 385
#define KW_TRY 386
#define KW_TYPEDEF 387
#define KW_TYPEID 388
#define KW_TYPENAME 389
#define KW_UNDERLYING_TYPE 390
#define KW_UNION 391
#define KW_UNSIGNED 392
#define KW_USING 393
#define KW_VIRTUAL 394
#define KW_VOID 395
#define KW_VOLATILE 396
#define KW_WCHAR_T 397
#define KW_WHILE 398
#define START_CPP 399
#define START_CONST_EXPR 400
#define START_TYPE 401



#if ! defined YYSTYPE && ! defined YYSTYPE_IS_DECLARED

# define yystype YYSTYPE /* obsolescent; will be withdrawn */
# define YYSTYPE_IS_DECLARED 1
#endif

#if ! defined YYLTYPE && ! defined YYLTYPE_IS_DECLARED
typedef struct YYLTYPE
{
  int first_line;
  int first_column;
  int last_line;
  int last_column;
} YYLTYPE;
# define yyltype YYLTYPE /* obsolescent; will be withdrawn */
# define YYLTYPE_IS_DECLARED 1
# define YYLTYPE_IS_TRIVIAL 1
#endif


#ifdef YYPARSE_PARAM
#if defined __STDC__ || defined __cplusplus
int cppyyparse (void *YYPARSE_PARAM);
#else
int cppyyparse ();
#endif
#else /* ! YYPARSE_PARAM */
#if defined __STDC__ || defined __cplusplus
int cppyyparse (void);
#else
int cppyyparse ();
#endif
#endif /* ! YYPARSE_PARAM */

#endif /* !YY_CPPYY_BUILT_TMP_CPPBISON_YXX_H_INCLUDED  */
